compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1591,6 +1590,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                         &self.misc(sp),
                         &mut |err| {
                             if let Some(expected_ty) = expected.only_has_type(self) {
+                                if blk.stmts.is_empty() && blk.expr.is_none() {
+                                    self.suggest_boxing_when_appropriate(
+                                        err,
+                                        blk.span,
+                                        blk.hir_id,
+                                        expected_ty,
+                                        self.tcx.mk_unit(),
+                                    );
```

Comment:
@compiler-errors this is why i split up the arguments, it was the only way i could figure out to pass both the `expr` and `blk` to `suggest_boxing_when_appropriate`, and the function only uses the span and hir_id anyways.